### PR TITLE
FileSystem extension skeleton

### PR DESCRIPTION
**Pull request by AFFogarty**
This PR provides the skeleton for an extension that implements `FileSystem`.  This PR only implements the `Delete()` API, which allows users to delete files.  The rest of the APIs are intentionally omitted so that they can be implemented in future PRs.

This PR also provides the skeleton for `Configuration`, which is returned by a new method `SparkContext.HadoopConfiguration()`.  This `Configuration` skeleton does not currently implement any methods, but can be passed to `FileSystem.Get()` to get the `FileSystem` for a given `SparkContext`.

Example:

```
SparkSession spark = ...

FileSystem fs = FileSystem.Get(spark.SparkContext.HadoopConfiguration())
```

This PR relates to #328.

### CODE REVIEW

File: src/csharp/Extensions/Microsoft.Spark.Extensions.FileSystem/FileSystem.cs [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+{
+    /// <summary>
+    /// An abstract base class for a fairly generic filesystem. It may be implemented as a distributed filesystem, or
+    /// as a "local" one that reflects the locally-connected disk. The local version exists for small Hadoop instances
+    /// and for testing.
+    /// 
+    /// All user code that may potentially use the Hadoop Distributed File System should be written to use an FileSystem
+    /// object. The Hadoop DFS is a multi-machine system that appears as a single disk.It's useful because of its fault
+    /// tolerance and potentially very large capacity.
+    /// </summary>
```
Comment (Niharikadutta) [inline src/csharp/Extensions/Microsoft.Spark.Extensions.FileSystem/FileSystem.cs:19]: Please reformat to keep each line within the 110 character limit

File: src/csharp/Extensions/Microsoft.Spark.Extensions.FileSystem/FileSystem.cs [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+using Microsoft.Spark.Interop.Ipc;
+
+namespace Microsoft.Spark.Extensions.FileSystem
+{
+    /// <summary>
+    /// An abstract base class for a fairly generic filesystem. It may be implemented as a distributed filesystem, or
+    /// as a "local" one that reflects the locally-connected disk. The local version exists for small Hadoop instances
+    /// and for testing.
+    /// 
+    /// All user code that may potentially use the Hadoop Distributed File System should be written to use an FileSystem
```
Comment (Niharikadutta) [inline src/csharp/Extensions/Microsoft.Spark.Extensions.FileSystem/FileSystem.cs:16]:
```suggestion
    /// All user code that may potentially use the Hadoop Distributed File System should be written to use a FileSystem
```

File: src/csharp/Extensions/Microsoft.Spark.Extensions.FileSystem/FileSystem.cs [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    /// object. The Hadoop DFS is a multi-machine system that appears as a single disk.It's useful because of its fault
```
Comment (Niharikadutta) [inline src/csharp/Extensions/Microsoft.Spark.Extensions.FileSystem/FileSystem.cs:17]:
```suggestion
    /// object. The Hadoop DFS is a multi-machine system that appears as a single disk. It's useful because of its fault
```

File: src/csharp/Extensions/Microsoft.Spark.Extensions.FileSystem/FileSystem.cs [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+        /// </summary>
+        /// <param name="sparkContext">The SparkContext whose configuration will be used.</param>
+        /// <returns>The FileSystem.</returns>
+        public static FileSystem Get(SparkContext sparkContext)
+        {
+            // TODO: Expose hadoopConfiguration as a .NET class and add an override for Get() that takes it.
+            JvmObjectReference hadoopConfiguration = (JvmObjectReference)
+                ((IJvmObjectReferenceProvider)sparkContext).Reference.Invoke("hadoopConfiguration");
+
+            return new JvmReferenceFileSystem(
```
Comment (Niharikadutta) [inline src/csharp/Extensions/Microsoft.Spark.Extensions.FileSystem/FileSystem.cs:33]: Why do we need `JvmReferenceFileSystem` to encapsulate the JVM object, why can't we do that within `FileSystem` itself?

File: src/csharp/Extensions/Microsoft.Spark.Extensions.FileSystem/FileSystem.cs [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+                    "org.apache.hadoop.fs.FileSystem",
+                    "get",
+                    hadoopConfiguration));
+        }
+
+        /// <summary>
+        /// Delete a file.
+        /// </summary>
+        /// <param name="path">The path to delete.</param>
+        /// <param name="recursive">If path is a directory and set to true, the directory is deleted else throws an
```
Comment (Niharikadutta) [inline src/csharp/Extensions/Microsoft.Spark.Extensions.FileSystem/FileSystem.cs:44]: Exceeding 110 character limit.

File: src/csharp/Extensions/Microsoft.Spark.Extensions.FileSystem/JvmReferenceFileSystem.cs [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+            _jvmObject = jvmObject;
+        }
+
+        JvmObjectReference IJvmObjectReferenceProvider.Reference => _jvmObject;
+
+        /// <summary>
+        /// Delete a file.
+        /// </summary>
+        /// <param name="path">The path to delete.</param>
+        /// <param name="recursive">If path is a directory and set to true, the directory is deleted else throws an
```
Comment (Niharikadutta) [inline src/csharp/Extensions/Microsoft.Spark.Extensions.FileSystem/JvmReferenceFileSystem.cs:28]: Exceeding 110 character limit

File: src/csharp/Extensions/Microsoft.Spark.Extensions.FileSystem/JvmReferenceFileSystem.cs [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+            JvmObjectReference pathObject =
+                SparkEnvironment.JvmBridge.CallConstructor("org.apache.hadoop.fs.Path", path);
+
+            return (bool)_jvmObject.Invoke("delete", pathObject, recursive);
+        }
+
+        public override void Dispose()
+        {
+            _jvmObject.Invoke("close");
+        }
```
Comment (Niharikadutta) [inline src/csharp/Extensions/Microsoft.Spark.Extensions.FileSystem/JvmReferenceFileSystem.cs:42]:
```suggestion
        public override void Dispose() => _jvmObject.Invoke("close");
```

File: src/csharp/Extensions/Microsoft.Spark.Extensions.FileSystem.E2ETest/Constants.cs [hunk 8]
```diff
@@ -0,0 +1,14 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+namespace Microsoft.Spark.Extensions.FileSystem.E2ETest
+{
+    /// <summary>
+    /// Constants related to the FileSystem test suite.
+    /// </summary>
+    internal class Constants
```
Comment (Niharikadutta) [inline src/csharp/Extensions/Microsoft.Spark.Extensions.FileSystem.E2ETest/Constants.cs:10]: Is this file needed? What else are we expecting to be added here?

File: src/csharp/Extensions/Microsoft.Spark.Extensions.FileSystem.E2ETest/FileSystemTests.cs [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+            using FileSystem fs = FileSystem.Get(_spark.SparkContext);
+
+            using var tempDirectory = new TemporaryDirectory();
+            string path = Path.Combine(tempDirectory.Path, "temp-table");
+            _spark.Range(25).Write().Format("parquet").Save(path);
+
+            Assert.True(fs.Delete(path, true));
+            Assert.False(fs.Delete(path, true));
+        }
+    }
```
Comment (Niharikadutta) [inline src/csharp/Extensions/Microsoft.Spark.Extensions.FileSystem.E2ETest/FileSystemTests.cs:38]: Can we add another test that validates the `Delete` API by checking if the file got deleted from the file system?

File: src/csharp/Extensions/Microsoft.Spark.Extensions.FileSystem.E2ETest/Microsoft.Spark.Extensions.FileSystem.E2ETest.csproj [hunk 10]
```diff
@@ -0,0 +1,13 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>netcoreapp3.1</TargetFramework>
+    <IsPackable>false</IsPackable>
```
Comment (Niharikadutta) [inline src/csharp/Extensions/Microsoft.Spark.Extensions.FileSystem.E2ETest/Microsoft.Spark.Extensions.FileSystem.E2ETest.csproj:5]: Why are we setting this to false?

File: src/csharp/Extensions/Microsoft.Spark.Extensions.FileSystem.E2ETest/Microsoft.Spark.Extensions.FileSystem.E2ETest.csproj [hunk 11]
```diff
@@ -0,0 +1,13 @@
...
+  </PropertyGroup>
+
+  <ItemGroup>
+    <ProjectReference Include="..\..\Microsoft.Spark.E2ETest\Microsoft.Spark.E2ETest.csproj" />
+    <ProjectReference Include="..\..\Microsoft.Spark\Microsoft.Spark.csproj" />
+    <ProjectReference Include="..\Microsoft.Spark.Extensions.FileSystem\Microsoft.Spark.Extensions.FileSystem.csproj" />
+  </ItemGroup>
```
Comment (Niharikadutta) [inline src/csharp/Extensions/Microsoft.Spark.Extensions.FileSystem.E2ETest/Microsoft.Spark.Extensions.FileSystem.E2ETest.csproj:12]: nit: add empty line

File: src/csharp/Extensions/Microsoft.Spark.Extensions.FileSystem/FileSystem.cs [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+    /// tolerance and potentially very large capacity.
+    /// </summary>
+    public abstract class FileSystem : IDisposable
+    {
+        /// <summary>
+        /// Returns the configured FileSystem implementation.
+        /// </summary>
+        /// <param name="sparkContext">The SparkContext whose configuration will be used.</param>
+        /// <returns>The FileSystem.</returns>
+        public static FileSystem Get(SparkContext sparkContext)
```
Comment (imback82) [inline src/csharp/Extensions/Microsoft.Spark.Extensions.FileSystem/FileSystem.cs:27]: This is weird. Is this like a factory? How can I create a new type of FileSystem?

Why not just expose the Hadoop FileSystem directly?

File: src/csharp/Extensions/Microsoft.Spark.Extensions.FileSystem/FileSystem.cs (hunk 4 above)
Comment (AFFogarty) [reply src/csharp/Extensions/Microsoft.Spark.Extensions.FileSystem/FileSystem.cs:33]: In the JVM implementation, `FileSystem` is an abstract class.  I wanted to keep that same pattern here.

File: src/csharp/Extensions/Microsoft.Spark.Extensions.FileSystem.E2ETest/Microsoft.Spark.Extensions.FileSystem.E2ETest.csproj (hunk 10 above)
Comment (AFFogarty) [reply src/csharp/Extensions/Microsoft.Spark.Extensions.FileSystem.E2ETest/Microsoft.Spark.Extensions.FileSystem.E2ETest.csproj:5]: Hmmm, I just copied this from the other extension `E2ETest` projects.  I will remove it everywhere.

File: src/csharp/Extensions/Microsoft.Spark.Extensions.FileSystem/FileSystem.cs (hunk 1 above)
Comment (AFFogarty) [reply src/csharp/Extensions/Microsoft.Spark.Extensions.FileSystem/FileSystem.cs]: Oops, I thought we increased it to 120 for some reason.  Fixed.

File: src/csharp/Extensions/Microsoft.Spark.Extensions.FileSystem/FileSystem.cs (hunk 5 above)
Comment (AFFogarty) [reply src/csharp/Extensions/Microsoft.Spark.Extensions.FileSystem/FileSystem.cs]: Fixed.

File: src/csharp/Extensions/Microsoft.Spark.Extensions.FileSystem.E2ETest/Constants.cs (hunk 8 above)
Comment (AFFogarty) [reply src/csharp/Extensions/Microsoft.Spark.Extensions.FileSystem.E2ETest/Constants.cs:10]: This is the pattern we're using in all the test projects.

File: src/csharp/Extensions/Microsoft.Spark.Extensions.FileSystem.E2ETest/FileSystemTests.cs (hunk 9 above)
Comment (AFFogarty) [reply src/csharp/Extensions/Microsoft.Spark.Extensions.FileSystem.E2ETest/FileSystemTests.cs:55]: Changed `TestSignatures()` to only test the signatures, and moved the functional testing into a new test that validates that the file is properly deleted.

File: src/csharp/Extensions/Microsoft.Spark.Extensions.FileSystem.E2ETest/Microsoft.Spark.Extensions.FileSystem.E2ETest.csproj (hunk 11 above)
Comment (AFFogarty) [reply src/csharp/Extensions/Microsoft.Spark.Extensions.FileSystem.E2ETest/Microsoft.Spark.Extensions.FileSystem.E2ETest.csproj:11]: Fixed.

File: src/csharp/Extensions/Microsoft.Spark.Extensions.FileSystem/FileSystem.cs (hunk 12 above)
Comment (AFFogarty) [reply src/csharp/Extensions/Microsoft.Spark.Extensions.FileSystem/FileSystem.cs:27]: This is the pattern from the Hadoop `FileSystem` class:  https://hadoop.apache.org/docs/current/api/org/apache/hadoop/fs/FileSystem.html#get-org.apache.hadoop.conf.Configuration-

FileSystem` is an abstract class with static `get()` factory methods that return concrete implementations based on the configuration parameters.

For my .NET implementation, I've added an override of `Get()` that takes `SparkContext` so that we don't have to expose `SparkContext.hadoopConfiguration`.

Comment (imback82) [reply src/csharp/Extensions/Microsoft.Spark.Extensions.FileSystem/FileSystem.cs:27]: If you are mimicking the hadoop `FileSystem`, shall we follow the same signature (expose minimal `Configuration`)? Also, we should add `Hadoop` to the namespace? (and let's add the link to the comment as well)

File: src/csharp/Extensions/Microsoft.Spark.Extensions.FileSystem/JvmReferenceFileSystem.cs [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+
+using Microsoft.Spark.Interop;
+using Microsoft.Spark.Interop.Ipc;
+
+namespace Microsoft.Spark.Extensions.FileSystem
+{
+    /// <summary>
+    /// <see cref="FileSystem"/> implementation that wraps a corresponding FileSystem object in the JVM.
+    /// </summary>
+    public class JvmReferenceFileSystem : FileSystem, IJvmObjectReferenceProvider
```
Comment (imback82) [inline src/csharp/Extensions/Microsoft.Spark.Extensions.FileSystem/JvmReferenceFileSystem.cs:13]: Why is `JvmReference` public?

Comment (imback82) [reply src/csharp/Extensions/Microsoft.Spark.Extensions.FileSystem/JvmReferenceFileSystem.cs:13]: I am thinking that we should just put APIs into `FileSystem` since `get` handles the getting the right concrete class. In what scenarios, do you see we need a concrete implementation of `FileSystem` other than hadoop filesystem?

Comment (imback82) [reply src/csharp/Extensions/Microsoft.Spark.Extensions.FileSystem/JvmReferenceFileSystem.cs:13]: And just name the package as `Microsoft.Spark.Extensions.Hadoop.FileSystem` so that there is no confusion; i.e., we are just wrapping `org.apache.hadoop.fs.FileSystem` more or less.

Comment (AFFogarty) [reply src/csharp/Extensions/Microsoft.Spark.Extensions.FileSystem/JvmReferenceFileSystem.cs]: Sure, I'll just make `FileSystem` concrete.  I was just trying to keep it as similar to the JVM implementation as possible.  But I suppose we won't have any POCO `FileSystem` implementations.

Comment (AFFogarty) [reply src/csharp/Extensions/Microsoft.Spark.Extensions.FileSystem/JvmReferenceFileSystem.cs]: Made `FileSystem` concrete and removed `JvmReferenceFileSystem`.

File: src/csharp/Extensions/Microsoft.Spark.Extensions.FileSystem/FileSystem.cs (hunk 12 above)
Comment (AFFogarty) [reply src/csharp/Extensions/Microsoft.Spark.Extensions.FileSystem/FileSystem.cs]: Added `Hadoop` to the namespace.

For this PR, I just wanted to provide an MVP skeleton so that it would be easy for community members to contribute APIs in additional PRs.  I'm thinking that we can invite others to contribute `Configuration` if they want to.  Thoughts?

Comment (imback82) [reply src/csharp/Extensions/Microsoft.Spark.Extensions.FileSystem/FileSystem.cs]: Looks like we cannot define `Configuration` class in the extension package, since `SparkContext.hadoopConfiguration` will be inside `Microsoft.Spark`..

What if we add `Hadoop` directory under https://github.com/dotnet/spark/tree/master/src/csharp/Microsoft.Spark and add `FileSystem.cs` and `Configuration.cs`? Note that we don't have to expose any of the APIs for `Configuration`.  I just want to be able to create `FileSystem` by `FileSystem.Get(sparkContext.HadoopConfiguration)`. Since we are at 1.0, we want to avoid breaking public APIs if possible.

Comment (AFFogarty) [reply src/csharp/Extensions/Microsoft.Spark.Extensions.FileSystem/FileSystem.cs]: Thoughts @rapoth? I know you wanted to put `FileSystem` in an extension.

File: src/csharp/Extensions/Microsoft.Spark.Extensions.Hadoop.FileSystem/FileSystem.cs [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+
+namespace Microsoft.Spark.Extensions.Hadoop.FileSystem
+{
+    /// <summary>
+    /// A fairly generic filesystem. It may be implemented as a distributed filesystem, or as a "local" one
+    /// that reflects the locally-connected disk. The local version exists for small Hadoop instances and for
+    /// testing.
+    /// 
+    /// All user code that may potentially use the Hadoop Distributed File System should be written to use a FileSystem
+    /// object. The Hadoop DFS is a multi-machine system that appears as a single disk. It's useful because of its fault
```
Comment (Niharikadutta) [inline src/csharp/Extensions/Microsoft.Spark.Extensions.Hadoop.FileSystem/FileSystem.cs:17]: Exceeding 110 character limit.

File: src/csharp/Extensions/Microsoft.Spark.Extensions.Hadoop.FileSystem/FileSystem.cs [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+            _jvmObject = jvmObject;
+        }
+
+        JvmObjectReference IJvmObjectReferenceProvider.Reference => _jvmObject;
+
+        /// <summary>
+        /// Returns the configured FileSystem implementation.
+        /// </summary>
+        /// <param name="sparkContext">The SparkContext whose configuration will be used.</param>
+        /// <returns>The FileSystem.</returns>
```
Comment (Niharikadutta) [inline src/csharp/Extensions/Microsoft.Spark.Extensions.Hadoop.FileSystem/FileSystem.cs:35]: nit: The FileSystem object.

**Review by Niharikadutta — COMMENTED** [review]
A few small nits but otherwise LGTM. Thanks @AFFogarty !

File: src/csharp/Extensions/Microsoft.Spark.Extensions.Hadoop.FileSystem/FileSystem.cs (hunk 15 above)
Comment (AFFogarty) [reply src/csharp/Extensions/Microsoft.Spark.Extensions.Hadoop.FileSystem/FileSystem.cs:35]: Made a reference.

File: src/csharp/Extensions/Microsoft.Spark.Extensions.Hadoop.FileSystem/FileSystem.cs (hunk 14 above)
Comment (AFFogarty) [reply src/csharp/Extensions/Microsoft.Spark.Extensions.Hadoop.FileSystem/FileSystem.cs]: Fixed.

File: src/csharp/Extensions/Microsoft.Spark.Extensions.FileSystem/FileSystem.cs (hunk 12 above)
Comment (AFFogarty) [reply src/csharp/Extensions/Microsoft.Spark.Extensions.FileSystem/FileSystem.cs]: Synced with @rapoth.  We will go with @imback82 's approach.

File: src/csharp/Extensions/Microsoft.Spark.Extensions.FileSystem/JvmReferenceFileSystem.cs (hunk 13 above)
Comment (AFFogarty) [reply src/csharp/Extensions/Microsoft.Spark.Extensions.FileSystem/JvmReferenceFileSystem.cs]: Changed implementation to use `Configuration`.

File: src/csharp/Microsoft.Spark/SparkContext.cs [hunk 16]
```diff
@@ -312,6 +313,13 @@ public Broadcast<T> Broadcast<T>(T value)
             return new Broadcast<T>(this, value);
         }
 
+        /// <summary>
+        /// A default Hadoop Configuration for the Hadoop code (e.g. file systems) that we reuse.
+        /// </summary>
+        /// <returns>The Hadoop Configuration.</returns>
+        public Configuration HadoopConfiguration() =>
```
Comment (AFFogarty) [inline src/csharp/Microsoft.Spark/SparkContext.cs:320]: @imback82 Not sure if this should be a function or a property.

Comment (imback82) [reply src/csharp/Microsoft.Spark/SparkContext.cs:320]: Since this is `def` on the Scala side, this will be a method on C# side. So the current approach is good.

File: src/csharp/Microsoft.Spark/Hadoop/FS/FileSystem.cs [hunk 17]
```diff
@@ -0,0 +1,58 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using Microsoft.Spark.Hadoop.Conf;
+using Microsoft.Spark.Interop;
+using Microsoft.Spark.Interop.Ipc;
+
+namespace Microsoft.Spark.Hadoop.FS
```
Comment (imback82) [inline src/csharp/Microsoft.Spark/Hadoop/FS/FileSystem.cs:10]:
```suggestion
namespace Microsoft.Spark.Hadoop.Fs
```
? Since we use `Ipc` instead of `IPC`?

File: src/csharp/Microsoft.Spark/Hadoop/FS/FileSystem.cs [hunk 18]
```diff
@@ -0,0 +1,58 @@
...
+        JvmObjectReference IJvmObjectReferenceProvider.Reference => _jvmObject;
+
+        /// <summary>
+        /// Returns the configured <see cref="FileSystem"/>.
+        /// </summary>
+        /// <param name="conf">The configuration to use.</param>
+        /// <returns>The FileSystem.</returns>
+        public static FileSystem Get(Configuration conf) =>
+            new FileSystem((JvmObjectReference)SparkEnvironment.JvmBridge.CallStaticJavaMethod(
+                "org.apache.hadoop.fs.FileSystem", "get", conf));
```
Comment (imback82) [inline src/csharp/Microsoft.Spark/Hadoop/FS/FileSystem.cs:39]: nit: I believe you need to break for each param.

File: src/csharp/Microsoft.Spark.E2ETest/Hadoop/FileSystemTests.cs [hunk 19]
```diff
@@ -0,0 +1,58 @@
...
+
+        /// <summary>
+        /// Test that methods return the expected signature.
+        /// </summary>
+        [Fact]
+        public void TestSignatures()
+        {
+            using var tempDirectory = new TemporaryDirectory();
+
+            using FileSystem fs = Assert.IsAssignableFrom<FileSystem>(
```
Comment (imback82) [inline src/csharp/Microsoft.Spark.E2ETest/Hadoop/FileSystemTests.cs:32]: you can remove `IsAssignableFrom` here since it's not going to compile if it doesn't return assignable `FileSystem`.

**Review by imback82 — COMMENTED** [review]
LGTM (few minor comments), thanks @AFFogarty!

**Comment by imback82** [comment]
@AFFogarty Can you also update the title/description now that this PR provides bindings to Hadoop Filesystem?

File: src/csharp/Microsoft.Spark.E2ETest/Hadoop/FileSystemTests.cs (hunk 19 above)
Comment (AFFogarty) [reply src/csharp/Microsoft.Spark.E2ETest/Hadoop/FileSystemTests.cs]: Removed.

File: src/csharp/TestApp/Program.cs [hunk 20]
```diff
@@ -0,0 +1,28 @@
+﻿using Microsoft.Spark.Extensions.Hadoop.FileSystem;
+using Microsoft.Spark.Sql;
+using System;
+
```
Comment (imback82) [inline src/csharp/TestApp/Program.cs:4]: Can we remove this file?

Comment (AFFogarty) [reply src/csharp/TestApp/Program.cs]: Oops, didn't mean to commit that. Removed.

File: src/csharp/Microsoft.Spark/Hadoop/FS/FileSystem.cs (hunk 18 above)
Comment (AFFogarty) [reply src/csharp/Microsoft.Spark/Hadoop/FS/FileSystem.cs]: Fixed

**Comment by AFFogarty** [comment]
Addressed nits and updated description.

**Comment by imback82** [comment]
@AFFogarty can you check the failed tests if it's transient?

**Comment by AFFogarty** [comment]
@imback82 I merged with `master` and now the tests are passing.

**Review by imback82 — APPROVED** [review]
LGTM, thanks @AFFogarty!
